lib/gosu/compat.rb

```diff
@@ -1,196 +1,48 @@
-
-    msg = if klass_or_full_message.is_a?(String) and name.nil? and repl.nil?
-      [ "DEPRECATION WARNING: #{klass_or_full_message},
-        \n#Called from #{caller[1, DEPRECATION_STACKTRACE_LINES]}"
-      ]
-    else
-      # Class method deprecation warnings result in something like this:
-      # #<Class:Gosu::Window>::button_id_to_char is deprecated
-      # Remove the instance-inspect stuff to make it look a bit better:
-      if klass_or_full_message.kind_of?(Module)
-        target = "#{klass_or_full_message.to_s.gsub(/#<Class:(.*)>/, '\1')}::"
-      else
-        "#{klass_or_full_message.class}#"
-      end
-      [ "DEPRECATION WARNING: #{target}#{name} is deprecated",
-        repl == :none ? " with no replacement." : "; use #{repl} instead.",
-        "\n#{target}#{name} called from #{Gosu.deprecation_caller.join("\n")}",
-      ]
+  # Support for KbLeft instead of KB_LEFT and Gp3Button2 instead of GP_3_BUTTON_2.
+  Gosu.constants.grep(/^KB_|MS_|GP_/).each do |new_name|
+    old_name = case new_name
+    when :KB_ISO then "KbISO"
+    when :KB_NUMPAD_PLUS then "KbNumpadAdd"
+    when :KB_NUMPAD_MINUS then "KbNumpadSubtract"
+    when :KB_EQUALS then "KbEqual"
+    when :KB_LEFT_BRACKET then "KbBracketLeft"
+    when :KB_RIGHT_BRACKET then "KbBracketRight"
+    else new_name.to_s.capitalize.gsub(/_(.)/) { $1.upcase }
     end
-    return if @@_deprecations_shown.has_key?(msg[0])
-    @@_deprecations_shown[msg[0]] = true
-
-    warn "#{msg.join}."
+    Gosu.const_set old_name, Gosu.const_get(new_name)
   end
 
-  # This method removes the deprecation methods themselves from the stacktrace.
-  def self.deprecation_caller
-    caller.delete_if { |trace_line| trace_line =~ /(deprecat|const_missing)/ }
-      .first(DEPRECATION_STACKTRACE_LINES)
+  def self.language
+    @language_cache ||= (user_languages.first || "en_US")
   end
-end
-
-# These were useful for working with Direct3D and the Win32 API a long time ago,
-# there was never a real reason to have them available in Ruby.
-Gosu.deprecate Gosu::Color, :bgr, :none
-Gosu.deprecate Gosu::Color, :abgr, :none
-
-# No need to pass a Window to Image.
-class Gosu::Image
-  alias_method :initialize_without_window, :initialize
 
-  def initialize(*args)
-    if args[0].is_a? Gosu::Window
-      Gosu.deprecation_message("Passing a Window to Image#initialize has been deprecated in Gosu 0.9 and this method now uses an options hash, see https://www.libgosu.org/rdoc/Gosu/Image.html ")
-      if args.size == 7
-        initialize_without_window args[1], tileable: args[2], rect: args[3..-1]
-      else
-        initialize_without_window args[1], tileable: args[2]
+  class Window
+    # Compat code taken from gosu/gosu
+
+    %w(update draw needs_redraw? needs_cursor? capture_cursor? hit_test gain_focus lose_focus
+      lose_focus button_down button_up gamepad_connected gamepad_disconnected drop close).each do |callback|
+      define_method "protected_#{callback}" do |*args|
```

Comment:
Gosu's SWIG wrapper used to call protected_foo from the C++ directory stuff. Whatever this is trying to do needs to be done in a different way now.

Reply:
Ah, window.rb still uses a similar mechanism even with FFI. Need to look deeper into this.